module/User/src/Listener/AuthenticationListener.php

```diff
@@ -0,0 +1,91 @@
+<?php
+
+namespace User\Listener;
+
+use InvalidArgumentException;
+use Laminas\Authentication\AuthenticationService;
+use Laminas\Mvc\MvcEvent;
+use Laminas\Stdlib\ResponseInterface as Response;
+use LogicException;
+use User\Adapter\ApiPrincipalAdapter;
+use User\Service\ApiAuthenticationService;
+
+final class AuthenticationListener
+{
+    // Defining the authentication types
+    public const AUTH_NONE = 'none';
+    public const AUTH_DBUSER = 'dbuser';
+    public const AUTH_API = 'api';
+
+    public function __construct(
+        private readonly AuthenticationService $authService,
+        private readonly ApiAuthenticationService $apiAuthService,
+        private readonly ApiPrincipalAdapter $apiPrincipalAdapter,
+    ) {
+    }
+
+    public function __invoke(MvcEvent $e)
+    {
+        if ($e->getName() !== MvcEvent::EVENT_ROUTE) {
+            throw new InvalidArgumentException(
+                "Expected MvcEvent of type " . MvcEvent::EVENT_ROUTE . ", got " . $e->getName()
+            );
+        }
+
+        $match = $e->getRouteMatch();
+        if ($match === null) {
+            throw new LogicException("Did not match any route after being routed");
+        }
+
+        switch ($match->getParam('auth_type', self::AUTH_DBUSER)) {
+            case self::AUTH_DBUSER:
+                return $this->dbuserauth($e);
+            case self::AUTH_API:
+                return $this->apiAuth($e);
+            case self::AUTH_NONE:
+                return;
+            default:
+                throw new InvalidArgumentException(
+                    "Authentication type was set to unknown type " . $match->getParam('auth_type')
+                );
+        }
+    }
+
+    /**
+     * Handle authentication for users
+     */
+    private function dbuserauth(MvcEvent $e): ?Response
```

Comment:
```diff
- private function dbuserauth(MvcEvent $e): ?Response
+ private function dbuserAuth(MvcEvent $e): ?Response
```

Reply:
Fixed in af35b1ea2eb2846bae0aa0da6fd79b8b83add97e